### PR TITLE
Make CTAS work.

### PR DESCRIPTION
**Pull request by guofengrichard**
The idea is to execute CTAS utility only on QD
and dispatch the Select part to QEs.

Diverging from PG, we add a flag in Query to indicate
if it is part of CTAS and add IntoClause in PlannedStmt
to dispatch target info to QEs.

Co-authored-by: Gang Xiong <gxiong@pivotal.io>

### CODE REVIEW

File: src/backend/parser/analyze.c [hunk 1]
```diff
@@ -3247,11 +3248,14 @@ transformCreateTableAsStmt(ParseState *pstate, CreateTableAsStmt *stmt)
...
-	((Query*)stmt->query)->intoClause = stmt->into;
+	/* GPDB: Set isCTAS to be true as we know this query is for CTAS */
+	((Query*)stmt->query)->isCTAS = true;
+
+	if (stmt->into->colNames)
+		applyColumnNames(((Query *)stmt->query)->targetList, stmt->into->colNames);
 
 	if (stmt->into->distributedBy && Gp_role == GP_ROLE_DISPATCH)
-		setQryDistributionPolicy((DistributedBy *) stmt->into->distributedBy, (Query *) stmt->query);
+		setQryDistributionPolicy((DistributedBy *) stmt->into->distributedBy, (Query *)stmt->query);
```
Comment (asimrp) [inline src/backend/parser/analyze.c]: Let's get rid of this unnecessary whitespace diff.

File: src/backend/parser/analyze.c [hunk 2]
```diff
@@ -3506,6 +3510,44 @@ applyLockingClause(Query *qry, Index rtindex,
 	qry->rowMarks = lappend(qry->rowMarks, rc);
 }
 
+/*
+ * Attach column names from a ColumnDef list to a TargetEntry list
```
Comment (asimrp) [inline src/backend/parser/analyze.c]: The function seems to assume that items in `ColumnDef` list and `TargetEntry` list are in the same order according to their attribute numbers.  Moreover, `TargetEntry` list may contain out-of-order junk attributes.  Is that assumption solid enough?  Is it possible to add assertions such as `(d->resorigcol == s->attnum)` to state the assumption more explicitly?

Comment (guofengrichard) [reply src/backend/parser/analyze.c]: This function is from PG 9.1 and has been removed from PG 9.2 and it is mis-used at this place. At least, `into->colNames` is a list of `T_String` rather than `ColumnDef`.
I refreshed the PR and removed this function. Thanks for reviewing.

File: src/include/executor/execdesc.h [hunk 3]
```diff
@@ -194,7 +194,7 @@ typedef struct QueryDispatchDesc
 	 * queryDesc->dest, use the original table's reloptions. If DestRemote is
 	 * set, use default reloptions + gp_default_storage_options.
 	 */
-	bool validate_reloptions;
+	bool useChangedAOOpts;
```
Comment (paul-guo-) [inline src/include/executor/execdesc.h:197]: I think we should not include AO in the variable name. That is too specific.

File: src/backend/executor/execMain.c [hunk 4]
```diff
@@ -677,9 +674,11 @@ standard_ExecutorStart(QueryDesc *queryDesc, int eflags)
 			dtmPreCommand("ExecutorStart", "(none)", queryDesc->plannedstmt,
 						  needDtxTwoPhase, true /* wantSnapshot */, queryDesc->extended_query );
 
-			queryDesc->ddesc->sliceTable = estate->es_sliceTable;
-
-			queryDesc->ddesc->oidAssignments = GetAssignedOidsForDispatch();
+			if (queryDesc->ddesc != NULL)
```
Comment (paul-guo-) [inline src/backend/executor/execMain.c:677]: no need of this line since previous code has checked ddesc as true.
